### PR TITLE
Dapp2 growth engine

### PR DESCRIPTION
**Pull request by sparrowDom**
### Description:
Implements active campaign page and part of enrolment modals flow. Currenlty the functionality is locked out with ENABLE_GROWTH environmental variable

**This Pull request: **
- implements resolvers to fetch campaign data from backend
- partly implements welcome and active campaign pages

**What is missing: **
- Enrolment modals don't yet enrol user to an active campaign
- campaign pagination
- referral page
- metamask enrolment signature
- polish

**To setup: **
- set `ENABLE_GROWTH` in `docker-compose.experimental.yml` to true. Or supply environmental variable differently
- visit `http://localhost:3001/#/welcome` page and `http://localhost:3001/#/campaigns` page

### Checklist:

- [ ] Test your work and double-check to confirm that you didn't break anything
- [ ] Wrap any displayed ETH addresses with [`formattedAddress`](https://github.com/OriginProtocol/origin/blob/master/origin-dapp/src/utils/user.js#L15-L17)
- [ ] Wrap any new text/strings for translation
- [ ] Run `npm run translations` if there are any changes to translated strings
- [ ] Map any new environment variables with a default value in the Webpack config
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/origin-docs)

### CODE REVIEW

File: experimental/origin-admin/src/queries/Fragments.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ export default {
   },
   Listing: {
     basic: gql`
-      fragment basicListingFields on ListingResult {
+      fragment basicListingFields on Listing {
```
Comment (franckc) [inline experimental/origin-admin/src/queries/Fragments.js]: Why are those changes part of this PR ? they seem unrelated...

File: experimental/origin-dapp2/src/css/app.css [hunk 2]
```diff
@@ -23,7 +23,7 @@ body {
 }
 
 .loading-screen {
-  background: url(images/spinner-animation-dark.svg) no-repeat center;
+  /* background: url('images/spinner-animation-dark.svg') no-repeat center;*/
```
Comment (franckc) [inline experimental/origin-dapp2/src/css/app.css]: remove ?

File: experimental/origin-dapp2/src/pages/growth/Welcome.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+          <div className="col-6 d-flex flex-column">
+            <Link to="/">
+              <img className="logo" src="/images/origin-logo-footer.svg" />
+            </Link>
+            <div className="title-text">
+              Your friend Aure has invited you to earn <b>FREE Origin Tokens</b>
+            </div>
+            <div className="sub-title-text">
+              Create an account on Origin today and start completing tasks for
+              the chance to will up to 2000 OGN currently valued at 2000 USD.
```
Comment (franckc) [inline experimental/origin-dapp2/src/pages/growth/Welcome.js]: nit: type. "the chance to earn up".

I assume this copy will get updated by Matt ? OGN is not valued at $1

File: experimental/origin-dapp2/src/queries/Fragments.js [hunk 4]
```diff
@@ -141,5 +142,39 @@ export default {
...
+        id
+        name
+        startDate
+        endDate
+        distributionDate
+        status
+        actions {
+          type
+          status
+          rewardEarned {
```
Comment (franckc) [inline experimental/origin-dapp2/src/queries/Fragments.js]: nit: Please check with Nick and Stan re standard for currency objects in GraphQL schema. There was a debate for flat vs nested fields. like rewardEarnedAmount and rewardEarnedCurrency vs rewardEnarned { amount, currency } and I don't remember what was decided :)

File: experimental/origin-graphql/src/typeDefs/Growth.js [hunk 5]
```diff
@@ -27,16 +27,28 @@ module.exports =
   }
 
   enum GrowthActionType {
-    basicProfile
-    attestations
+    email
+    phoneNumber
```
Comment (franckc) [inline experimental/origin-graphql/src/typeDefs/Growth.js]: supernit. maybe we could just call this one phone.

File: experimental/origin-graphql/src/typeDefs/Growth.js [hunk 6]
```diff
@@ -27,16 +27,28 @@ module.exports =
...
+    listingPurchased
   }
 
   enum GrowthInviteStatus {
     pending
     successful
   }
 
+  enum Eligibility {
+    Eligible
```
Comment (franckc) [inline experimental/origin-graphql/src/typeDefs/Growth.js]: supernit: shoudl this be lowercap, similarily to the other enums above ?

File: origin-discovery/src/listener/listener.js [hunk 7]
```diff
@@ -304,6 +304,8 @@ const config = {
   defaultContinueBlock: parseInt(process.env.CONTINUE_BLOCK || 0)
 }
 
+logger.debug("CONFIG", JSON.stringify(config))
```
Comment (franckc) [inline origin-discovery/src/listener/listener.js]: remove since config is already logged below at line 348 ?

File: origin-growth/src/apollo/app.js [hunk 8]
```diff
@@ -37,7 +37,22 @@ const server = new ApolloServer({
...
-  playground: true
+  playground: true,
+  context: context => {
+    let userIp = null
+    const headers = context.req.headers
+    /* TODO: this needs to be tested on production. On localhost x-forwarded-for headers are not set
+     * - mark this as a Pull Request check list item and then delete it
+     */
+    if (headers) {
+      userIp = headers['x-forwarded-for'] || null
```
Comment (franckc) [inline origin-growth/src/apollo/app.js]: We run in Google Cloud which adds the header "x-appengine-user-ip" from which you can grab the user's IP.

File: origin-growth/src/apollo/resolvers.js [hunk 9]
```diff
@@ -10,15 +12,63 @@ const resolvers = {
...
+        return {
+          code: 200,
+          success: true,
+          eligibility: 'Eligible',
+          countryName: 'N/A',
+          countryCode: 'N/A'
+        }
+      }
+
+      const locationInfo = await getLocationInfo(context.userIp)
```
Comment (franckc) [inline origin-growth/src/apollo/resolvers.js]: Note that in the back-end we could simply retrieve the location info directly from the headers that google cloud automatically adds to any request (see this [doc](https://cloud.google.com/appengine/docs/standard/go/reference/request-response-headers) and look for header x-appengine-country .

It would be faster and we would not have to worry about requests quota / paying for the service.. Though downside is that it could be slightly different compared to any location info done by the front-end (unless if switching front-end to use Google Cloud Function geolocation that I had exposed as an experiment).

File: origin-growth/src/enums.js [hunk 10]
```diff
@@ -20,6 +20,32 @@ const GrowthCampaignRewardStatuses = new Enum(
   'Distributed'
 )
 
+const GrowthCampaignStatuses = new Enum(
```
Comment (franckc) [inline origin-growth/src/enums.js]: nit: how should we capitalize those enum values ? I used upper case. But maybe I should have used lower case ? I don't really care one way or the other but maybe we should try to be consistent.

File: origin-growth/src/rules/rules.js [hunk 11]
```diff
@@ -1,11 +1,59 @@
...
+    if (first.currency !== second.currency)
+      throw new Error(`At least two rewards have different currencies. ${first.currency} ${second.currency}`)
+    return {
+      amount: BigNumber(first.amount).plus(BigNumber(second.amount)),
+      currency: first.currency
+    }
+  })
+
+  return {
+    amount: finalReward.amount.toString(),
```
Comment (franckc) [inline origin-growth/src/rules/rules.js]: Note that all rules amount are in natural unit so the sum will be in natural unit as well.
e.g. for 1 OGN, as opposed to showing 1000000000000000 it would return 1.

Should those be converted to token unit here or will you be handling this at a different layer ?

File: origin-growth/src/rules/rules.js [hunk 12]
```diff
@@ -1,11 +1,59 @@
...
 
 // System cap for number of rewards per rule.
 const MAX_NUM_REWARDS_PER_RULE = 1000
 
+const sumUpRewards = (rewards) => {
+  if (rewards === null || rewards.length === 0){
+    return null
+  }
+
+  const finalReward = rewards.reduce((first, second) => {
```
Comment (franckc) [inline origin-growth/src/rules/rules.js]: super super nit: maybe call totalReward ?

File: origin-growth/src/rules/rules.js [hunk 13]
```diff
@@ -1,11 +1,59 @@
...
+  })
+
+  return {
+    amount: finalReward.amount.toString(),
+    currency: finalReward.currency
+  }
+}
+
+const eventTypeToActionType = (eventType) => {
+  if (eventType === 'ProfilePublished'){
```
Comment (franckc) [inline origin-growth/src/rules/rules.js]: alternative to this long list of if/else statements: do a lookup in a dictionary.

eventTypeToActionTypeMap = {
 ProfilePublished: 'profile',
 EmailAttestationPublished: 'email'
 ...
}
return eventTypeToActionTypeMap[eventType]

File: origin-growth/src/rules/rules.js [hunk 14]
```diff
@@ -139,7 +235,7 @@ class Level {
   qualifyForNextLevel(ethAddress, events) {
     for (let i = 0; i < this.rules.length; i++) {
       const result = this.rules[i].qualifyForNextLevel(ethAddress, events)
-      if (result != null && result === false) {
+      if (result !== null && result === false) {
```
Comment (franckc) [inline origin-growth/src/rules/rules.js]: good catch ! ty !

File: origin-growth/src/util/fixtures.js [hunk 15]
```diff
@@ -19,15 +19,44 @@ function tokenNaturalUnits(x) {
...
+            visible: true
+          },
+        },
+        {
+          id: 'EmailAttestation',
+          class: 'SingleEvent',
+          config: {
+            eventType: 'EmailAttestationPublished',
+            reward: {
+              amount: tokenNaturalUnits(10),
```
Comment (franckc) [inline origin-growth/src/util/fixtures.js]: In the [spec](https://docs.google.com/document/d/1DJ00rl1MYSxkH5UMjVD0V-kzWU2uyRnAGA8x2YPqQdE/edit), it does not say email attestation will be rewarded - it is just used to qualify user for next level.

File: experimental/origin-admin/src/queries/Fragments.js (hunk 1 above)
Comment (sparrowDom) [reply experimental/origin-admin/src/queries/Fragments.js]: tnx for catching this, yes not part of this PR

File: experimental/origin-dapp2/src/css/app.css (hunk 2 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/css/app.css]: I had to comment this out because it was (weirdly) causing some build errors. It doesn't anymore so thanks removing this comment

File: experimental/origin-dapp2/src/pages/growth/Welcome.js (hunk 3 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/growth/Welcome.js]: thanks, fixing. Yup the copys are still to be completed

File: experimental/origin-dapp2/src/queries/Fragments.js (hunk 4 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/queries/Fragments.js]: will do 👍

File: experimental/origin-graphql/src/typeDefs/Growth.js (hunk 5 above)
Comment (sparrowDom) [reply experimental/origin-graphql/src/typeDefs/Growth.js]: yup better, changing

File: experimental/origin-graphql/src/typeDefs/Growth.js (hunk 6 above)
Comment (sparrowDom) [reply experimental/origin-graphql/src/typeDefs/Growth.js]: good point, changing to lower case

File: origin-discovery/src/listener/listener.js (hunk 7 above)
Comment (sparrowDom) [reply origin-discovery/src/listener/listener.js]: Yeah this has no place in here :) deleting

File: origin-growth/src/apollo/app.js (hunk 8 above)
Comment (sparrowDom) [reply origin-growth/src/apollo/app.js]: Great, grabbing ip from there. Good to know Google does heavy lifting of resolving IP

File: origin-growth/src/apollo/resolvers.js (hunk 9 above)
Comment (sparrowDom) [reply origin-growth/src/apollo/resolvers.js]: Super helpful! This even removes the need of obtaining the IP address. The only downside is that we can only test this once it is deployed. I guess this will be a thing to test on staging. I will add location data into /#/app-info so we can check there what country has been detected

File: origin-growth/src/enums.js (hunk 10 above)
Comment (sparrowDom) [reply origin-growth/src/enums.js]: Yeah I noticed that. I used lower case because these are directly the values used in `Growth.js` GraphQL typedef. I guess we could: 
1. make these upper case and call `toLowerCase()` before returning values in the resolvers. 
2. make these lowercase across the system. 

Do you have any preference? Perhaps 1. would be easier to implement

File: origin-growth/src/rules/rules.js (hunk 12 above)
Comment (sparrowDom) [reply origin-growth/src/rules/rules.js]: sure 👍 better

File: origin-growth/src/rules/rules.js (hunk 11 above)
Comment (sparrowDom) [reply origin-growth/src/rules/rules.js]: Currently the frontend receives natural units and queries the wallet for decimal places of the token's currency. Then converts to token units. I would keep it this way?

File: origin-growth/src/rules/rules.js (hunk 13 above)
Comment (sparrowDom) [reply origin-growth/src/rules/rules.js]: Nice, that is much cleaner... 👍

File: origin-growth/src/rules/rules.js (hunk 14 above)
Comment (sparrowDom) [reply origin-growth/src/rules/rules.js]: 👍

File: origin-growth/src/util/fixtures.js (hunk 15 above)
Comment (sparrowDom) [reply origin-growth/src/util/fixtures.js]: good catch, changing to null

File: origin-growth/src/rules/rules.js (hunk 11 above)
Comment (franckc) [reply origin-growth/src/rules/rules.js]: Yes, I think what you have is better. Let's keep it this way.

File: origin-growth/src/enums.js (hunk 10 above)
Comment (franckc) [reply origin-growth/src/enums.js]: I don't have any strong preference... We haven't run the DB schema in prod for growth so there is no operational overhead if we decide to go for option 2.
Do what you think is best :)

**Comment by sparrowDom** [comment]
@nick when you have time can you CR this please? The whole growth flow is not yet fully implemented, but It would be nice if we can merge this part so it doesn't become an even bigger pull request. 

Also a question:  currently in GrowthSchema we use an object to represent price fields (object consisting of `currency` and `amount` fields. Should that be flattened out into 2 fields or left as an object?

When glancing through code you can also estimate how this will affect the stability of dapp2.

File: experimental/origin-admin/src/queries/UserOffers.js [hunk 16]
```diff
@@ -1,5 +1,4 @@
 import gql from 'graphql-tag'
-import fragments from './Fragments'
```
Comment (nick) [inline experimental/origin-admin/src/queries/UserOffers.js]: Did we need to get rid of this?

File: experimental/origin-dapp2/src/pages/App.js [hunk 17]
```diff
@@ -1,4 +1,4 @@
-import React, { Component } from 'react'
+import React, { Component, Fragment } from 'react'
```
Comment (nick) [inline experimental/origin-dapp2/src/pages/App.js]: We can use `<></>` instead of `<Fragment></Fragment>` now

File: experimental/origin-dapp2/src/pages/App.js [hunk 18]
```diff
@@ -50,11 +52,18 @@ class App extends Component {
         </div>
       )
     }
+
+    const shouldRenderNavbar = () => {
```
Comment (nick) [inline experimental/origin-dapp2/src/pages/App.js]: can this just be a variable instead of a function? like `const shouldRenderNavbar = this.props.location.pathname !== '/welcome'`

File: experimental/origin-dapp2/src/pages/growth/Campaigns.js [hunk 19]
```diff
@@ -0,0 +1,615 @@
...
+  let { type, status, reward, rewardEarned, rewardPending } = props.action
+  const showLockIcon = status === 'inactive'
+
+  const actionCompleted = ['exhausted', 'completed'].includes(status)
+  let backgroundImgSrc = actionCompleted
+    ? 'images/identity/verification-shape-green.svg'
+    : 'images/identity/verification-shape-blue.svg'
+
+  const formatTokens = tokenAmount => {
+    return web3.utils
```
Comment (nick) [inline experimental/origin-dapp2/src/pages/growth/Campaigns.js]: We should try to avoid any web3 calls in dapp2 and rely on this coming from GraphQL instead

File: experimental/origin-dapp2/src/pages/growth/Campaigns.js [hunk 20]
```diff
@@ -0,0 +1,615 @@
...
+      </div>
+    </Fragment>
+  )
+}
+
+function Campaign(props) {
+  const { campaign, accountId } = props
+  const { startDate, endDate, status, rewardEarned, actions, name } = campaign
+
+  const formatTimeDifference = (start, end) => {
```
Comment (nick) [inline experimental/origin-dapp2/src/pages/growth/Campaigns.js]: Perhaps we can move this to `utils/formatTimeDifference` - there's already a `utils/distanceToNow.js`

File: experimental/origin-admin/src/queries/UserOffers.js (hunk 16 above)
Comment (sparrowDom) [reply experimental/origin-admin/src/queries/UserOffers.js]: No fragments are used in this file. Should we still keep it in?

File: experimental/origin-dapp2/src/pages/App.js (hunk 17 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/App.js]: Yup I've seen that and it is pretty handy. Somehow sublime babel syntax plugin isn't able to handle this and for that reason I didn't use it. Can't find an elegant solution for this

File: experimental/origin-dapp2/src/pages/App.js (hunk 18 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/App.js]: Uff I over complicated this one... thanks!

File: experimental/origin-dapp2/src/pages/growth/Campaigns.js (hunk 19 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/growth/Campaigns.js]: Got the idea from here: https://github.com/OriginProtocol/origin/blob/master/experimental/origin-dapp2/src/hoc/withTokenBalance.js#L26

😈 

Will resolve by adding bignumber.js npm package.

File: experimental/origin-dapp2/src/pages/growth/Campaigns.js (hunk 20 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/growth/Campaigns.js]: Good idea, doing that tnx

**Comment by sparrowDom** [comment]
@nick or @franckc all the comments have been addressed. Can I get a stamp? Or if you want you can do another pass OFC.

File: experimental/origin-admin/src/queries/UserOffers.js (hunk 16 above)
Comment (nick) [reply experimental/origin-admin/src/queries/UserOffers.js]: There used to be a fragment on line 29... any reason it was removed?

File: experimental/origin-dapp2/package.json [hunk 21]
```diff
@@ -35,6 +35,7 @@
     "apollo-link": "^1.2.8",
     "babel-plugin-module-resolver": "^3.2.0",
     "blueimp-load-image": "^2.20.1",
+    "big-number": "^2.0.0",
```
Comment (nick) [inline experimental/origin-dapp2/package.json]: Sorry to be a pain... looks like web3 is in fact used throughout dapp2 haha. Perhaps we can remove this dependency and go back to the old way of using web3.utils, for consistency?

File: experimental/origin-dapp2/src/pages/App.js [hunk 22]
```diff
@@ -83,6 +85,20 @@ class App extends Component {
             <Route path="/notifications" component={Notifications} />
             <Route path="/about/dapp-info" component={DappInfo} />
             <Route path="/about/tokens" component={AboutToken} />
+            <Route
```
Comment (nick) [inline experimental/origin-dapp2/src/pages/App.js]: can we have a `const GrowthEnabled = process.env.ENABLE_GROWTH === 'true'` at the top and then `{GrowthEnabled && <Route path="/about/tokens" component={GrowthCampaigns} />}` here?

File: experimental/origin-dapp2/src/pages/about/DappInfo.js [hunk 23]
```diff
@@ -52,7 +53,35 @@ const DappInfo = () => (
...
+              {({ networkStatus, error, loading, data }) => {
+                if (networkStatus === 1 || loading) {
+                  return dataTr({
+                    key: 'Detected country',
+                    value: 'Loading...'
+                  })
+                } else if (error) {
+                  return dataTr({
+                    key: 'Detected country',
+                    value: 'Error getching country'
```
Comment (nick) [inline experimental/origin-dapp2/src/pages/about/DappInfo.js]: fetching 😇

File: experimental/origin-dapp2/src/queries/Fragments.js (hunk 4 above)
Comment (nick) [reply experimental/origin-dapp2/src/queries/Fragments.js]: I think we decided to flatten it but we haven't gotten around to it yet...

File: experimental/origin-graphql/src/mutations/web3/signMessage.js [hunk 24]
```diff
@@ -0,0 +1,6 @@
+import contracts from '../../contracts'
+
+export default async (_, { address, message }) => {
+  //TODO: get this right
```
Comment (nick) [inline experimental/origin-graphql/src/mutations/web3/signMessage.js]: is there an issue with this?

File: experimental/origin-dapp2/src/pages/about/DappInfo.js (hunk 23 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/about/DappInfo.js]: haa :D what a typo

File: experimental/origin-admin/src/queries/UserOffers.js (hunk 16 above)
Comment (sparrowDom) [reply experimental/origin-admin/src/queries/UserOffers.js]: I might be missing something here. Does a fragment definition serve any purpose if it is not referenced inside the query? 
Example: 
```
nodes {
  id
  listingId
  ... basicListingFields
}
```

Since there was no `...basicListingFields` I though I should just remove it?

File: experimental/origin-dapp2/package.json (hunk 21 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/package.json]: Sure no problem :)

File: experimental/origin-admin/src/queries/UserOffers.js (hunk 16 above)
Comment (nick) [reply experimental/origin-admin/src/queries/UserOffers.js]: Ahhh yea you're right, my bad :)

File: experimental/origin-dapp2/src/pages/App.js (hunk 22 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/pages/App.js]: 👍

File: experimental/origin-dapp2/src/queries/Fragments.js (hunk 4 above)
Comment (sparrowDom) [reply experimental/origin-dapp2/src/queries/Fragments.js]: Ok leaving it as it is for now then

File: experimental/origin-graphql/src/mutations/web3/signMessage.js (hunk 24 above)
Comment (sparrowDom) [reply experimental/origin-graphql/src/mutations/web3/signMessage.js]: It just worked without Metamask prompting user to sign (I was expecting a popup window). Was planning to get to the bottom of this in the next PR.

**Comment by sparrowDom** [comment]
@nick ready for another look

File: origin-growth/src/rules/rules.js [hunk 25]
```diff
@@ -123,6 +169,57 @@ class Campaign {
     }
     return rewards
   }
+
+  /**
```
Comment (franckc) [inline origin-growth/src/rules/rules.js]: These methods are not really related to the rules calculation.
So maybe we should put them under src/resources/campaign.js ?
We can discuss. This does not have to be done in this PR.

File: origin-growth/src/util/locationInfo.js [hunk 26]
```diff
@@ -0,0 +1,269 @@
+const restrictedCountryCodes = ['US', 'CA']
+const forbiddenCountryCodes = []
+
+const countryMap = {
+  AD: 'Andorra',
```
Comment (franckc) [inline origin-growth/src/util/locationInfo.js]: We'll need to figure out how to localize those strings...
Maybe add a TODO ?

File: origin-growth/src/rules/rules.js (hunk 25 above)
Comment (sparrowDom) [reply origin-growth/src/rules/rules.js]: Cool have added an item to [our document](https://docs.google.com/document/d/15nzudW56hQVIWHc7DRJI5XzxenM0-oc2p7z4vOqc8ms/edit?ts=5b7b483f#heading=h.b3i2k22a9u94)

File: origin-growth/src/util/locationInfo.js (hunk 26 above)
Comment (sparrowDom) [reply origin-growth/src/util/locationInfo.js]: I see there are some public repositories of country names in each of the languages. Adding a TODO. 

You think we should have this implemented for initial launch of growth engine?

Comment (franckc) [reply origin-growth/src/util/locationInfo.js]: I think it is ok to punt on this for the launch and come back to it a bit later....
